x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityIndexManager.java

```diff
@@ -350,53 +308,6 @@ public void prepareIndexIfNeededThenExecute(final Consumer<Exception> consumer,
             } else if (indexState.indexExists() && indexState.isIndexUpToDate == false) {
                 throw new IllegalStateException("Index [" + indexState.concreteIndexName + "] is not on the current version."
                         + "Security features relying on the index will not be available until the upgrade API is run on the index");
-            } else if (indexState.indexExists() == false) {
```

Comment:
I think we still need to handle the case where the master node doesn't know about the security system index, so for BWC we should maintain this code somehow.

Reply:
Interesting point.
I think we presently don't handle the cases of index auto-creation and mapping updates too well.
if `.security` does not yet exist, rolling update is in progress, and an API that creates the index for the first time hits a new node, the index will be created, but the other old nodes will subsequently complain that the index is not up to date (because the format number is greater) or that the mapping is not up to date (because the mapping version is greater). 

I think this might be a common issue, I vaguely remember something similar for async search. Would it be possible to defer the system index upgrade (of the mapping and of the settings/metadata) until the rolling upgrade is complete, and let the _security business logic_ deal with possibly storing entities in the old format, or if that's not possible, return a response failure informing that the requested API (with its parameters) is not available in a mixed cluster (similar to how no APIs are available until the cluster state has been recovered)?

Reply:
I've reinstated the code for creating the security index, while changing the code so that it relies on the auto-create logic. I've also re-added code for the mappings not being up-to-date, where the runnable is ignore and the exception consumer is called. Is that what you had in mind?

I've also changed a number of locations to check for a different in minimum and maximum ES version across the cluster nodes. If there is a different, the auto-create logic will refuse to run, and the `SystemIndexManager` will not try to update mappings.

Reply:
> I've reinstated the code for creating the security index

I'm not sure this is need, the auto-create should handle it (now that auto-create has a predictable behaviour in a mixed cluster).

> I've also re-added code for the mappings not being up-to-date, where the runnable is ignore and the exception consumer is called. Is that what you had in mind?

This fixes the problem, yes.

> I've also changed a number of locations to check for a different in minimum and maximum ES version across the cluster nodes.

This is very cool. I like it that there is a clear, testable behavior in a mixed cluster scenario.

Reply:
Hmm, if the create call is kept around, the rename from `prepareIndexIfNeededThenExecute`  to `checkIndexStateThenExecute` is less fortunate 👿